src/function/mtk.jl

```diff
@@ -58,7 +60,19 @@ function instantiate_function(f, x, adtype::AutoModelingToolkit, p, num_cons=0)
         cons_h = f.cons_h
     end
 
-    return OptimizationFunction{true}(f.f, adtype; grad=grad, hess=hess, hv=hv, 
+    if adtype.obj_sparse
+        _hess_prototype = ModelingToolkit.hessian_sparsity(sys)
+        hess_prototype = sparse(findnz(_hess_prototype)[1], findnz(_hess_prototype)[2], zeros(nnz(_hess_prototype)))
+    end
+
+    if adtype.cons_sparse
+        _cons_jac_prototype = ModelingToolkit.jacobian_sparsity(cons_sys)
+        cons_jac_prototype = sparse(findnz(_cons_jac_prototype)[1], findnz(_cons_jac_prototype)[2], zeros(nnz(_cons_jac_prototype)))
+        _cons_hess_prototype = ModelingToolkit.hessian_sparsity(cons_sys)
+        cons_hess_prototype = [sparse(findnz(_cons_hess_prototype[i])[1], findnz(_cons_hess_prototype[i])[2], zeros(nnz(_cons_hess_prototype[i]))) for i in eachindex(_cons_hess_prototype)]
```

Comment:
Simplify these too.

Reply:
I was doing the conversion at the solver before but then decided to store the correctly typed prototype to use directly, it seemed cleaner than doing it for different solvers (already doing it at two seemed like unnecessary repetition of code)?

Reply:
No, since you want to make sure it's thread-safe, so you have to do a copy operation at the solver anyways, so you might as well do the type promotion there too. Trying to correctly type it here is also wrong because you don't necessarily have sufficient information: if someone does a remake to `Dual` numbers this will be incorrectly typed and error. Those are the two reasons it's done at the solver level.